### PR TITLE
[train] Fix issues in migration of tune_cifar_torch_pbt_example

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Resolves three issues that come up when migrating the `tune_cifar_torch_pbt_example` from Ray 2.6 to Ray 2.7:

1. There is a warning message because PBT uses the `_schedule_trial_save` interface. This is added to the white list attributes so it doesn't come up anymore.
2. PBT malfunctions in Python 2.7, so instead of silently failing, we raise an error and ask users to migrate
3. When users use old `ray.air.Checkpoint` APIs on `ray.train.Checkpoint`, we should raise an actionable error message.

For 3) a bit more background, and @pcmoritz and @ericl:

We currently raise an error in the new storage mode:

```
                raise ValueError(
                    "You must pass a `ray.train.Checkpoint` "
                    "object to `train.report`. `ray.air.Checkpoint` is deprecated."
                )
```

This suggests users to switch over, but it does not suggest the API surface is different. A natural thing to do is just to switch the import statements. In this case, if `from_dict/to_dict` etc is called, I think it's sensible to raise an actionable error message with migration instructions.

Using a metaclass has the advantage of not having to add the deprecated methods to the new class.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/schedulers/pbt.py [hunk 1]
```diff
@@ -361,6 +361,17 @@ def __init__(
         require_attrs: bool = True,
         synch: bool = False,
     ):
+        if not _use_storage_context():
+            raise RuntimeError(
+                "Due to breaking API changes, PBT does not work with the old "
+                "persistence mode (enabled via RAY_AIR_NEW_PERSISTENCE_MODE=0). "
+                "Migrate your script to use the new APIs instead and disabled the "
```
Comment (justinvyu) [inline python/ray/tune/schedulers/pbt.py:368]: I think we should just not mention this env var -- all user code even w/o PBT will already error, so we should just tell them: Please migrate to the new APIs to use PBT. 

And link to this issue: `https://github.com/ray-project/ray/issues/37868`

File: python/ray/train/_checkpoint.py [hunk 2]
```diff
@@ -26,8 +26,25 @@
...
+        try:
+            return super().__getattribute__(item)
+        except AttributeError as exc:
+            if item in [
+                "from_dict",
+                "to_dict",
+                "from_bytes",
+                "to_bytes",
+                "from_uri",
+                "to_uri",
```
Comment (justinvyu) [inline python/ray/train/_checkpoint.py:40]: Can we also include: `uri`, `get_internal_representation`, `get_preprocessor` and `set_preprocessor`

Comment (justinvyu) [reply python/ray/train/_checkpoint.py:40]: `uri`, `from_uri` and `to_uri` should have a different message:

To create a checkpoint from remote storage, create a Checkpoint using its constructor instead of `from_directory`. Ex: `Checkpoint(path="s3://a/b/c")`. Then, access the contents of the checkpoint with `as_directory` / `to_directory`. The `ray.train.Checkpoint` no longer provides an API to upload to remote storage using a checkpoint -- use the 
`pyarrow.fs.FileSystem` attribute (`checkpoint.filesystem`) or the  client of your choice to do this instead.

Comment (justinvyu) [reply python/ray/train/_checkpoint.py:40]: `get/set_preprocessor` should have a different error message:

To include preprocessor information in checkpoints, pass it as `metadata` in the `<Framework>Trainer` constructor. Ex: `TorchTrainer(..., metadata={...})`.  After training, access it in the checkpoint via `checkpoint.get_metadata()`.

Comment (justinvyu) [reply python/ray/train/_checkpoint.py:40]: get internal representation can keep the same error message as the current ones

File: python/ray/train/_checkpoint.py [hunk 3]
```diff
@@ -301,3 +318,24 @@ def _list_existing_del_locks(path: str) -> List[str]:
...
+        f"with tempfile.TemporaryDirectory() as checkpoint_dir:\n"
+        f"  with open(os.path.join(checkpoint_dir, 'data.json'), 'wb') as fp:\n"
+        f"    pickle.dump({{'data': 'value'}}, fp)\n\n"
+        f"  checkpoint = Checkpoint.from_directory(checkpoint_dir)\n"
+        f"  train.report(..., checkpoint=checkpoint)\n\n"
+        f"Example to load a dictionary from a checkpoint:\n\n"
+        f"if train.get_checkpoint():\n"
+        f"  with train.get_checkpoint().as_directory() as checkpoint_dir:\n"
+        f"    with open(os.path.join(checkpoint_dir, 'data.json'), 'rb') as fp:\n"
+        f"      data = pickle.load(fp)"
```
Comment (justinvyu) [inline python/ray/train/_checkpoint.py:340]: json.load/dump

**Review by justinvyu — COMMENTED** [review]
Thanks! had some suggestions for more granular error messages for different deprecated methods.

File: python/ray/train/_checkpoint.py (hunk 3 above)
Comment (krfricke) [reply python/ray/train/_checkpoint.py:340]: ah I think I'll just use `data.pkl` instead. I think json may be too restrictive

File: python/ray/tune/schedulers/pbt.py (hunk 1 above)
Comment (krfricke) [reply python/ray/tune/schedulers/pbt.py:368]: SGTM

Comment (krfricke) [reply python/ray/tune/schedulers/pbt.py:368]: Hm I guess this error message only comes up when someone actually used that var?

File: python/ray/train/_checkpoint.py (hunk 2 above)
Comment (krfricke) [reply python/ray/train/_checkpoint.py:40]: Good suggestion, updated!

**Review by ericl — APPROVED** [review]
LGTM, approved.

Just out of curiosity, why a metaclass instead of defining getattr directly?

**Comment by krfricke** [comment]
> LGTM, approved.
> 
> Just out of curiosity, why a metaclass instead of defining getattr directly?

Because it's a `@classmethod` - `getattr` only works on instances, but since we use `from_bytes` etc as a constructor, the object doesn't exist, yet, and getattr on the derived class is not invoked.
